SAMZA-2222: Documentation for async API

mynameborat opened the pull request:
![image](https://user-images.githubusercontent.com/46942335/58350496-e8b5f880-7e1a-11e9-8736-16f968146109.png)

prateekm commented on docs/learn/tutorials/versioned/samza-async-user-guide.md:
```diff
@@ -107,16 +107,96 @@ task.callback.timeout.ms=5000
 
+### Asynchronous Process in High Level API
+
+If your application is asynchronous, e.g. making non-blocking remote IO calls, [AsyncFlatMapFunction](/learn/documentation/{{site.version}}/api/javadocs/org/apache/samza/operators/functions/AsyncFlatMapFunction.html) provides support for it. The following example illustrates an application that processes Wikipedia feed updates and invokes a remote service to standardize the updates and sends the  ...
```
"If your processing logic is asynchronous, e.g, it makes non-blocking remote calls, you can implement it using the AsyncFlatMapFunction".

prateekm commented on docs/learn/tutorials/versioned/samza-async-user-guide.md:
```diff
@@ -107,16 +107,96 @@ task.callback.timeout.ms=5000
+}
+{% endhighlight %}
+
+In the above example, the messages are not sent to Wikipedia after `AsyncStandardizerFunction` returns. The framework keeps track of the future returned by the function and propagates the standardized result to downstream operator (in this case sendTo) only when the future completes. Samza has a timeout for the future to complete and can be configured through the following property:
```
"The framework keeps track ..." etc is implementation detail. User guide should only talk about what the semantics are. E.g. , "In the above example, the results from the asynchronous ... function are propagated to downstream operators once the future is complete."

"Samza has a timeout for the future ..." makes it sound like we control when the future is completed or have an await. It'll be better to just say that "there is an overall timeout for each message to be processed and the processAsync callback to be invoked, and you can tune that using ..."

mynameborat replied:
leaving out "the processAsync callback to be invoked" since it feels implementation detail too.

prateekm commented on docs/learn/tutorials/versioned/samza-async-user-guide.md:
```diff
@@ -107,16 +107,96 @@ task.callback.timeout.ms=5000
+In the above example, the messages are not sent to Wikipedia after `AsyncStandardizerFunction` returns. The framework keeps track of the future returned by the function and propagates the standardized result to downstream operator (in this case sendTo) only when the future completes. Samza has a timeout for the future to complete and can be configured through the following property:
+
+{% highlight jproperties %}
+# Timeout for the future to complete. When the timeout elapses, the framework will throw a TaskCallbackTimeoutException and shut down the container.
```
"will throw a ..." is implementation detail. Also, can just say "the container will be restarted".

prateekm commented on docs/learn/tutorials/versioned/samza-async-user-guide.md:
```diff
@@ -107,16 +107,96 @@ task.callback.timeout.ms=5000
+task.callback.timeout.ms
+{% endhighlight %}
+
+For remote IO libraries that don't support Futures/Promises and use callbacks, it should be straightforward to adapt the callbacks to Future pattern. The following code snippet illustrates it
```
"If the I/O library accepts callbacks instead of returning a Future, the callback can be adapted to a Future in the following way: "

prateekm commented on docs/learn/tutorials/versioned/samza-async-user-guide.md:
```diff
@@ -107,16 +107,96 @@ task.callback.timeout.ms=5000
+{% highlight java %}
+
+  public CompletionStage<Collection<StandardizedWikipediaFeedEvent>> apply(WikipediaFeedEvent wikipediaFeedEvent) {
+    Request<StandardizerRequest> standardizedRequest = buildStandardizedRequest(wikipediaFeedEvent);
```
s/standardizedRequest/standardizationRequest

prateekm commented on docs/learn/tutorials/versioned/samza-async-user-guide.md:
```diff
@@ -107,16 +107,96 @@ task.callback.timeout.ms=5000
 {% endhighlight %}
 
-**NOTE:** In case of AsyncStreamTask, processAsync() is still invoked in the order of the message arrivals, but the completion can go out of order. In case of StreamTask with multithreading, process() can run out-of-order since they are dispatched to a thread pool. This option should **NOT** be used when strict ordering of the output is required.
+**NOTE:** In case of AsyncStreamTask, processAsync() is still invoked in the order of the message arrivals, but the completion can go out of order. In case of StreamTask and StreamApplication with concurrency, process() can run out-of-order since they are dispatched to a thread pool. This option should **NOT** be used when strict ordering of the output is required.
```
s/can go/can happen.

s/StreamApplication with concurrency/High Level API applications with task.max.concurrency > 1,

s/process() can run/delivery can be.

"dispatched to a thread pool" isn't relevant here.

mynameborat commented:
@prateekm Addressed your comments. Some of the language used initially was done with the notion of keeping up consistency within the document. 
The page still has some implementation detail and we may want to make it consistent. Will file a follow up JIRA.
